### PR TITLE
Fix the JAX `RandomVariable` dispatcher

### PR DESCRIPTION
**Pull request by rlouf**
The JAX dispatcher for `RandomVariable` `Op`s is currently broken, testing is lacking and the shapes are not handled at all. 

- [x] Refactor the dispatcher to handle the API differences in JAX;
- [x] Handle the shape when specified as a constant, or the output of a `Shape` operator (two cases were the shape is a `ConcreteArray`); fail gracefully otherwise;
- [x] Register an implementation for `StudentTRV`
- [x] Add comprehensive test suite
    - [x] Every implementation needs to be tested;
    - [x] Add test where the `size` parameter is the (dynamic) shape of a tensor.
    - [x] Add test with implementation for custom `RandomVariable` definition 
- [x] Register implementation for `GumbelRV`
- [x] Register implementation for `StandardNormalRV`
- [x] Open an issue for each missing `RandomVariable` implementation
- [x] Rewrite `MakeVector` and `DimShuffle` when passed as `size` parameter

### CODE REVIEW

**Comment by rlouf** [comment]
I have now registered the JAX implementation for every Aesara `RandomVariable` that has a JAX implementation in the JAX library individually, and added some minimal test for each.

I will now try to generalize the implementation, and improve the tests / make them more exhaustive.

**Comment by rlouf** [comment]
The transpilation process of `RandomVariable`s now goes throught two functions:

- The first, registered for the generic `RandomVariable` handles the `size` parameters; if we know at compile-time it is a constant use this constant, if it is the output of a `Shape` operator then use the output of a previous node. These are the two situations in which JAX will not complain, so we fail gracefully otherwise.
- The second, `jax_sample_fn` registers the implementation of the sampling function for each `RandomVariable` subtype  individually. This way users can register an implementation for the `RandomVariable`s they defined without having to worry about JAX's quirks.

We should use this pattern throughout the JAX backend to handle the `shape` arguments, which is one big pain point with this backend. Along with powerful shape inference on our side it should be enough to handle most problematic cases gracefully. 

There's a third case I did not consider for now, since it involves deeper changes in the backend, but that we should eventually support:

```python
import aesara
import aesara.tensor as at

size = at.iscalar()
x_rv = at.random.normal(0, 1, size=size)

fn = aesara.function([size], x_rv)
```

In this case, JAX will happily JIT-compile the function if `size` is in the `static_argnums` parameters. We would need a simple system that tracks the parameters that should be passed as static to `jax.jit`, and use this information at the end of the transpilation process when the function is jit-compiled. Alternatively we can defer jit-compilation to the user and let them set this parameter as static should they want to. The choice here matters, because JAX will re-compile the function each time a new value is passed for parameters marked as static.

**Comment by rlouf** [comment]
> I am not sure what MakeVector is used for besides this case, but we may need to use a different Op here on the JAX side to differentiate between cases.

Unless we choose to represent tuples in Aesara to represent the `size` and `shape` parameters I will need to indeed create a dummy `ShapeTuple` `Op` and rewrite the graph before transpiling.

**Comment by rlouf** [comment]
This is not perfect, but I found a happy middle ground by walking through the size node's ancestors until a `Shape` node is found, raising if the traversed nodes are not `MakeVector` or `Subtensor` or if no `Shape` node is found.

**Comment by rlouf** [comment]
I added a rewrite that replaces `MakeVector` and `DimShuffle` (when used to convert a scalar into a 1d vector) by a dummy `JAXShapeTuple` `Op` when found as a `size` input to a `RandomVariable`. This is my first time writing rewrites, and I need help to both make this work, and better organize the code.

### How to make this work?

The most pressing problem I have is that the linker does not seem to transpile the new `JAXShapeTuple` (even though I registered it). On the following example:

```python
import aesara.tensor as at

rng = shared(np.random.RandomState(123))
x_at = at.dmatrix()
out = at.random.normal(0, 1, size=(x_at.shape[0],), rng=rng)
jax_fn = function([x_at], out, mode=jax_mode)
assert jax_fn(np.ones((2, 3))).shape == (2,)
```

The `FunctionGraph` instance passed to `jax_funcify` in `aesara.link.jax.linker` does contain the `JAXShapeTuple` `Op` instead of the original `MakeVector` `Op`:

```python
# normal_rv{0, (0, 0), floatX, False}.1 [id A] 3
#  |RandomStateSharedVariable(<RandomState(MT19937) at 0x7FDD0F5C7140>) [id B]
#  |JAXShapeTuple [id C] 2
#  | |Subtensor{int64} [id D] 1
#  |   |Shape [id E] 0
#  |   | |<TensorType(float64, (?, ?))> [id F]
#  |   |ScalarConstant{0} [id G]
#  |TensorConstant{11} [id H]
#  |TensorConstant{0} [id I]
#  |TensorConstant{1} [id J]

```

But it seems to be skipped during transpilation: 

```python
  def jax_funcified_fgraph(tensor_variable, random_state_shared_variable):
      # Shape(<TensorType(float64, (?, ?))>)
      tensor_variable_1 = shape(tensor_variable)
      # Subtensor{int64}(Shape.0, ScalarConstant{0})
      tensor_variable_2 = subtensor(tensor_variable_1, scalar_constant)
      # MakeVector{dtype='int64'}(Subtensor{int64}.0)
      tensor_variable_3 = makevector(tensor_variable_2)
      # normal_rv{0, (0, 0), floatX, False}(RandomStateSharedVariable(<RandomState(MT19937) at 0x7F3BE3FC7140>), JAXShapeTuple.0, TensorConstant{11}, TensorConstant{0}, TensorConstant{1})
      variable, tensor_variable_5 = sample_fn(random_state_shared_variable, tensor_variable_4, tensor_constant, tensor_constant_1, tensor_constant_2)
```

### How to better organize the code?

The rewrite currently lives in `aesara.link.jax.linker` as a standalone rewrite, but this is obviously not ideal. What is the recommended way to store and apply the rewrites here?

File: aesara/link/jax/linker.py [hunk 1]
```diff
@@ -49,6 +49,34 @@ def fgraph_convert(self, fgraph, input_storage, storage_map, **kwargs):
...
+                    # Here Aesara is turning a scalar input into a 1d vector
+                    if (
+                        size_arg.owner.op.input_broadcastable == ()
+                        and size_arg.owner.op.new_order == ("x",)
+                    ):
+                        new_size_args = JAXShapeTuple()(*size_arg.owner.inputs)
+                        node.inputs[1] = new_size_args
+                        return node.outputs
+
+        simplify = WalkingGraphRewriter(size_parameter_as_tuple)
```
[the remark this answers is not in the document]
Comment (rlouf) [reply aesara/link/jax/linker.py]: I added the rewrite in a standalone `aesara.link.jax.dispatch.rewriting` module, registered it in `optdb` under the "jax" tag there, included the "jax" tag in the database query, but I don't know how to make the registration in `optdb` effective (`mode.__position__` currently does not contain the rewrite).

**Review by brandonwillard — COMMENTED** [review]
I pushed some changes that should allow the new JAX-specific rewrite to run in `"JAX"` mode.  

The import order/circularity was a challenge that was addressed by moving the rewrite into a sub-package for the relevant rewrites (i.e. `aesara.tensor.random` rewrites).  In this case, the import question shifts toward the use of `JAXShapeTuple`, which is defined in a module that requires an import of `jax`, but that was answered with a "lazy" import inside the rewrite itself.  This doesn't seem like the best approach, but it's not particularly bad; regardless, we should revisit this situation, because I can see us adding more backend-specific rewrite customizations like this, and we should have a good approach for that sooner than later.

For instance, we could define `JAXShapeTuple` somewhere else, but the question is "Where?".  Anyway, we can still move forward as is.

**Review by brandonwillard — APPROVED** [review]

**Pull request merged**